### ISSUE TITLE
Questions about the paper.

### ISSUE THREAD

**Issue by fancyerii** [Dao-AILab/flash-attention#618]
I have read [ELI5: FlashAttention](https://gordicaleksa.medium.com/eli5-flash-attention-5c44017022ad) and the author has some questions about the original paper.
1. As for B_r, I’m not exactly sure why do they perform a min op with d?
    I think we should make sure 4 x M/4d x B_r <= M because block of Q,K,V and S need be loaded to GPU SRAM. Am I right? If it's true, why let B_r = B_c = sqrt(M/4)? So 4  x B_r x B_c = M.
2. How do l_i & m_i fit into the SRAM (including all of the intermediate variables) when we computed block size in such a way that we only have enough space for K_j, V_j, Q_i & O_i? I think the answer is: registers.
   Did the author guess correctly?
3. How do they get to that number? Well, let’s analyze the nested for loop:

    Our block size is M/4d. This means that the vectors are split into N/(M/4d) blocks.
    Raise that to the power of 2 (because we’re looping across row/column blocks) and we get O(N² d² / M²)
    Now here I have M² and they only have M¹ — my hypothesis is that we probably need M/4d memory accesses inside of the loop to actually fetch all the vectors, i.e. we can’t fetch the whole block all at once. That would explain the missing M? (d can be ignored as a constant in big O).

  I think we should calculate as N² d² / M²(visit time) x M( memory/time) = N² d² / M. Am I right?

**Comment by tridao** [comment]
> 1. As for B_r, I’m not exactly sure why do they perform a min op with d?
>    I think we should make sure 4 x M/4d x B_r <= M because block of Q,K,V and S need be loaded to GPU SRAM. Am I right? If it's true, why let B_r = B_c = sqrt(M/4)? So 4  x B_r x B_c = M.

Right, you want blocks of Q, K, V, and S to fit in SRAM. I think B_c is set to M/4d and S has size B_r x B_c so we don't want B_r to be larger than 4d, otherwise S wouldn't fit. I think I just set B_c <= d to be safe so that S doesn't take more than 1/4 of the SRAM. You can also set B_r and B_c differently as long as these four matrices fit.
In practice these are set by what's easy to implement (e.g. powers of 2 like 64 are 128 are much easier to implement) and what's fast. 

> 2. How do l_i & m_i fit into the SRAM (including all of the intermediate variables) when we computed block size in such a way that we only have enough space for K_j, V_j, Q_i & O_i? I think the answer is: registers.
>    Did the author guess correctly?

Yes they're in registers. But they're also pretty small (size B_r I think).

> 3. How do they get to that number? Well, let’s analyze the nested for loop:
>    Our block size is M/4d. This means that the vectors are split into N/(M/4d) blocks.
>    Raise that to the power of 2 (because we’re looping across row/column blocks) and we get O(N² d² / M²)
>    Now here I have M² and they only have M¹ — my hypothesis is that we probably need M/4d memory accesses inside of the loop to actually fetch all the vectors, i.e. we can’t fetch the whole block all at once. That would explain the missing M? (d can be ignored as a constant in big O).
> 
> I think we should calculate as N² d² / M²(visit time) x M( memory/time) = N² d² / M. Am I right?

You can see the proof of theorem 1 in Appendix C.

**Comment by fancyerii** [comment]
thank you very much.
